### ISSUE TITLE
default backend is not deterministic

### ISSUE THREAD

**Issue by syrkuit** [jaraco/keyring#372]
When no backends are available, keyring.backends.chainer.ChainerBackend & keyring.backends.fail.Keyring end up having the same priority of 0
Because keyring.backend.get_all_keyring() is not deterministic, keyring.init_backend() isn't either as shown below.
As a result, behaviour is unpredictable, including raising a vague RuntimeError() exception:

```
$ repeat 10 python3 -c 'import keyring
print(keyring.backend.get_all_keyring()) 
try:                                                 
  print(keyring.get_password("not", "deterministic"))
except Exception as e:
  print(str(e))'
[<keyring.backends.chainer.ChainerBackend object at 0x7f914dcd85c0>, <keyring.backends.fail.Keyring object at 0x7f914ce7ad30>]
None
[<keyring.backends.fail.Keyring object at 0x7f8ed963d5f8>, <keyring.backends.chainer.ChainerBackend object at 0x7f8ed87dfc88>]
No recommended backend was available. Install the keyrings.alt package if you want to use the non-recommended backends. See README.rst for details.
[....]
```

**Comment by syrkuit** [comment]
btw- this is Python 3.7.2, with debian python3-keyring 17.1.1-1

**Comment by jaraco** [comment]
Indeed, that's undesirable. Thanks for the report.

**Comment by jaraco** [comment]
One way to address this would be to set the priority lower than 0 on the chainer, though that doesn't feel like a robust approach.

**Comment by syrkuit** [comment]
Yeah, that seemed like an easy way out, but not particularly clean which is why I didn't bother with a PR :-)

**Comment by remd** [comment]
I am experiencing different behavior. In a shell for-loop I always get `keyring.backends.fail.Keyring` and in a Python REPL I always get `keyring.backends.chainer.ChainerBackend`. Perhaps this is not surprising if the behavior is not deterministic. 

I am using Python 3.6.7 and keyring 18.0.0 from pip. 

In repeat loop:
```
$ for i in {1..10}; do python -c 'import keyring
print(keyring.backend.get_all_keyring())
try:
  print(keyring.get_password("not", "deterministic"))
except Exception as e:
  print(str(e))'; done
[<keyring.backends.fail.Keyring object at 0x7f76466ed940>, <keyring.backends.chainer.ChainerBackend object at 0x7f76432b97f0>]
No recommended backend was available. Install a recommended 3rd party backend package; or, install the keyrings.alt package if you want to use the non-recommended backends. See https://pypi.org/project/keyring for details.
[<keyring.backends.fail.Keyring object at 0x7f698e78d940>, <keyring.backends.chainer.ChainerBackend object at 0x7f698b359748>]
No recommended backend was available. Install a recommended 3rd party backend package; or, install the keyrings.alt package if you want to use the non-recommended backends. See https://pypi.org/project/keyring for details.
[<keyring.backends.fail.Keyring object at 0x7fa2ca4dd940>, <keyring.backends.chainer.ChainerBackend object at 0x7fa2c70a97f0>]
No recommended backend was available. Install a recommended 3rd party backend package; or, install the keyrings.alt package if you want to use the non-recommended backends. See https://pypi.org/project/keyring for details.
[....]
```

In Python REPL:
```
$ python
Python 3.6.7 (default, Oct 22 2018, 11:32:17)
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import keyring
>>> print(keyring.backend.get_all_keyring())
[<keyring.backends.chainer.ChainerBackend object at 0x7fbc3c402a58>, <keyring.backends.fail.Keyring object at 0x7fbc38951b00>]
$ python
Python 3.6.7 (default, Oct 22 2018, 11:32:17)
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import keyring
>>> print(keyring.backend.get_all_keyring())
[<keyring.backends.chainer.ChainerBackend object at 0x7f04d99a1a58>, <keyring.backends.fail.Keyring object at 0x7f04d5ef0b00>]
```

**Comment by earonesty** [comment]
Maybe sort them alphabetically, assuming at least some people will choose bad priorities for backends some of the time.

**Comment by ddstreet** [comment]
Can you clarify what the intended behavior is when there are no backends?  The behavior of an empty backends.chainer instance is to silently ignore all stored credentials, and return None for all requests (technically all its functions just fall off the end since the for loop never executes); while the behavior of a backends.fail instance is to raise RuntimeError for all its functions.

If there are no backends, it seems like the silent no-op behavior of the empty backends.chainer is better than killing the calling program with a RuntimeError, isn't it?

If backends.fail really should be used, can a new defined error be added like NoKeyringError or something, so that callers can at least sanely handle the situation?  I think in most cases it's likely preferable to simply proceed without any keyring to store in, than to crash.

**Comment by ddstreet** [comment]
@jaraco I opened PR #424 for this bug, can you review/merge?

**Comment by jaraco** [comment]
Fixed in #424 and released in v21.2.0.
